v3/lint/configuration.go

```diff
@@ -0,0 +1,248 @@
+// is responsible for closing the reader, if appropriate.
+func NewConfig(r io.Reader) (Configuration, error) {
+	tree, err := toml.LoadReader(r)
+	if err != nil {
+		return Configuration{}, err
+	}
+	return Configuration{tree}, nil
+}
+
+// NewConfigFromFile attempts to instantiate a configuration from the provided filesystem path.
+//
+// The file pointed to by `path` MUST be valid TOML file. If `path` is the empty string then
+// an empty configuration is returned.
+func NewConfigFromFile(path string) (Configuration, error) {
+	if path == "" {
+		return NewEmptyConfig(), nil
+	}
+	f, err := os.Open(path)
+	if err != nil {
+		return Configuration{}, fmt.Errorf("failed to open the provided configuration at %s. Error: %s", path, err.Error())
+	}
+	defer f.Close()
+	return NewConfig(f)
+}
+
+// NewConfigFromString attempts to instantiate a configuration from the provided string.
+//
+// The provided string MUST be in a valid TOML format.
+func NewConfigFromString(config string) (Configuration, error) {
+	return NewConfig(strings.NewReader(config))
+}
+
+// NewEmptyConfig returns a configuration that is backed by an entirely empty TOML tree.
+//
+// This is useful if no particular configuration is set at all by the user of ZLint as
+// any attempt to resolve a namespace in `deserializeConfigInto` fails and thus results
+// in all defaults for all lints being maintained.
+func NewEmptyConfig() Configuration {
+	ctx, _ := NewConfigFromString("")
+	return ctx
+}
+
+// deserializeConfigInto deserializes the section labeled by the provided `namespace`
+// into the provided target `interface{}`.
+//
+// For example, given the following configuration...
+//
+// ```
+// [e_some_lint]
+// field = 1
+// flag = false
+//
+// [w_some_other_lint]
+// is_web_pki = true
+// ```
+//
+// And the following struct definition...
+//
+// ```
+// type SomeOtherLint {
+//		IsWebPKI bool `toml:"is_web_pki"`
+// }
+// ```
+//
+// Then the invocation of this function should be...
+//
+// ```
+// lint := &SomeOtherLink{}
+// deserializeConfigInto(link, "w_some_other_lint")
+// ```
+//
+// If there is no such namespace found in this configuration then provided the namespace specific data encoded
+// within `target` is left unmodified. However, configuration of higher scoped fields will still be attempted.
+func (c Configuration) deserializeConfigInto(target interface{}, namespace string) error {
+	if tree := c.tree.Get(namespace); tree != nil {
+		err := tree.(*toml.Tree).Unmarshal(target)
+		if err != nil {
+			return err
+		}
+	}
+	return c.resolveHigherScopedReferences(target)
+}
+
+// resolveHigherScopeReferences takes in an interface{} value and attempts to
+// find any field within its inner value that is either a struct or a pointer
+// to a struct that is one of our global configurable types. If such a field
+// exists then that higher scoped configuration will be copied into the value
+// held by the provided interface{}.
+//
+// This procedure is recursive.
```

Comment:
You are correct that there is a caching opportunity here, however I was apprehensive when building this in providing lints with shared data for which I do not believe I can reasonably control the mutability.

```go
struct A {
    Config *RFC5280Config
}

func (a *A) Execute(cert *x509.Certificate) {
    A.Config.Field = "now something else entirely"
}

struct A {
    Config *RFC5280Config
}

func (b *B) Execute(cert *x509.Certificate) {
    // I sure hope that no one modified my config before I ran because that would be super sneaky
}
```

I suppose that for the common configurations I could do a series of `getters`, although I cannot do that in general for all configurations as I cannot know their beforehand.

However, a quick measurement shows that this is not much a problem (at least not at the moment).

```go
func BenchmarkConfig(b *testing.B) {
	config, err := lint.NewConfigFromString(`
[e_ca_common_name_missing2]
BeerHall = "liedershousenssss"
`)
	if err != nil {
		b.Fatal(err)
	}
	lint := NewCaCommonNameMissing()
	for i := 0; i < b.N; i++ {
		err = config.Configure(lint, "e_ca_common_name_missing2")
		if err != nil {
			b.Fatal(err)
		}
	}
}
```

...yields...

```
goos: linux
goarch: amd64
pkg: github.com/zmap/zlint/v3/test
cpu: Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
BenchmarkConfig
BenchmarkConfig-8   	  931068	      1250 ns/op
PASS
```

And indeed, simply tight looping this reconstruction 100k times take about 0.13s.